File: App/Resources/AlgorithmDetails/binaryinsertionsort.bundle/binaryinsertionsort.c
#include <stdio.h>
#include <stdlib.h>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23,
                 90, 69, 51, 81, 68, 83, 32, 56};

void printList(int arr[], int n) {
  for (int i = 0; i < n; i++) {
    if (i == 0) {
      printf("[%d, ", arr[i]);
    } else if (i != n - 1) {
      printf("%d, ", arr[i]);
    } else {
      printf("%d]", arr[i]);
    }
  }
}

int binarySearch(int arr[], int item, int start, int end) {
  int low = start;
  int high = end;
  while (low < high) {
    int mid = low + (high - low) / 2;
    if (item < arr[mid]) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }
  return low;
}

void sort(int arr[], int n) {
  for (int i = 1; i < n; i++) {
    int item = arr[i];
    int pos = binarySearch(arr, item, 0, i);
    int j = i;
    while (j > pos) {
      arr[j] = arr[j - 1];
      j--;
    }
    arr[pos] = item;
  }
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}

File: App/Resources/AlgorithmDetails/binaryinsertionsort/binaryinsertionsort.rb
def binary_search(array, item, first, last)
  low = first
  high = last
  while low < high
    mid = low + (high - low) / 2
    if item < array[mid]
      high = mid
    else
      low = mid + 1
    end
  end
  low
end

def binary_insertion_sort(array)
  (1...array.length).each do |i|
    item = array[i]
    pos = binary_search(array, item, 0, i)
    j = i
    while j > pos
      array[j] = array[j - 1]
      j -= 1
    end
    array[pos] = item
  end
  array
end

def sort(array)
  binary_insertion_sort(array)
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array
